File: 프로그래머스/lv1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.java
class Solution {
    public String solution(int[] numbers, String hand) {
        String answer = "";
        int[] left = {3, 0};
        int[] right = {3, 2};
        for(int i = 0; i < numbers.length; i++) {
        	int x = numbers[i] == 0? 3 : (numbers[i]-1)/3;
        	int y = numbers[i] == 0? 1 : (numbers[i]+2)%3; 
        	
        	if(getSuitableFinger(x, y, left, right, numbers[i], hand).equals("left")) {
        		answer += "L";
        		left[0] = x;
        		left[1] = y;
        	} else {
        		answer += "R";
        		right[0] = x;
        		right[1] = y;
        	}
        	
        }
        return answer;
    }
    
    public String getSuitableFinger(int x, int y, int[] left, int[] right, int number, String hand) {
		if(number % 3 == 1) {
    		return "left";
    	} else if(number % 3 == 0 && number != 0) {
    		return "right";
    		
    	} else if(Math.abs(left[0]-x) + Math.abs(left[1]-y) < Math.abs(right[0]-x) + Math.abs(right[1]-y)) {
    		return "left";
    		
    	} else if(Math.abs(left[0]-x) + Math.abs(left[1]-y) > Math.abs(right[0]-x) + Math.abs(right[1]-y)) {
    		return "right";
    		
    	} else {
    		return hand;
    	}
	}
}

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.js
function position(num){
    if (num === '*') num = '10';
    if (num === '0') num = '11';
    if (num === '#') num = '12';
    let number = parseInt(num);
    return [parseInt((number - 1) / 3), (number - 1) % 3];
}

function distance(finger, target){
    let fingerLocation = position(finger);
    let targetLocation = position(target);
    return Math.abs(fingerLocation[0] - targetLocation[0]) + 
                     Math.abs(fingerLocation[1] - targetLocation[1]);
}

function solution(numbers, hand) {
    let answer = '';
    let left = '*';
    let right = '#';
    for (let i = 0; i < numbers.length; i++){
        if (numbers[i] === 1 || numbers[i] === 4 || numbers[i] === 7){
            answer += 'L';
            left = numbers[i] + '';
        } else if (numbers[i] === 3 || numbers[i] === 6 || numbers[i] === 9){
            answer += 'R';
            right = numbers[i] + '';
        } else {
            if (distance(left, numbers[i] + '') < distance(right, numbers[i] + '')){
                answer += 'L';
                left = numbers[i] + '';
            } else if (distance(left, numbers[i] + '') > distance(right, numbers[i] + '')){
                answer += 'R';
                right = numbers[i] + '';
            } else {
                if (hand === 'right'){
                    answer += 'R';
                    right = numbers[i] + '';
                } else {
                    answer += 'L';
                    left = numbers[i] + '';
                }
            }
        }
    }
    return answer;
}